cmake/Modules/G4BuildSettings.cmake

```diff
@@ -153,75 +157,6 @@ endif() #NOT WIN32
 # The following options affect the compilation of the Geant4 libraries
 #
 
-#.rst
-# - ``GEANT4_BUILD_CXXSTD`` (Allowed values: 11, 14, 17, c++11, c++14, c++17)
-#
-#   - Choose C++ Standard to build against from supported list.
-#   - Note that only C++17 is supported on Windows with MSVC
-#
-set(__g4_default_cxxstd 11 14 17 c++11 c++14 c++17)
-if(MSVC)
-  set(__g4_default_cxxstd 17 c++17)
-endif()
-
-enum_option(GEANT4_BUILD_CXXSTD
-  DOC "C++ Standard to compile against"
-  VALUES ${__g4_default_cxxstd}
-  CASE_INSENSITIVE
-  )
-
-string(REGEX REPLACE "^c\\+\\+" "" GEANT4_BUILD_CXXSTD "${GEANT4_BUILD_CXXSTD}")
-mark_as_advanced(GEANT4_BUILD_CXXSTD)
-geant4_add_feature(GEANT4_BUILD_CXXSTD "Compiling against C++ Standard '${GEANT4_BUILD_CXXSTD}'")
-
-
-# Require at least C++11 with no extensions and the following features
-set(CMAKE_CXX_EXTENSIONS OFF)
-set(GEANT4_TARGET_COMPILE_FEATURES
-  cxx_alias_templates
-  cxx_auto_type
-  cxx_delegating_constructors
-  cxx_enum_forward_declarations
-  cxx_explicit_conversions
-  cxx_final
-  cxx_lambdas
-  cxx_nullptr
-  cxx_override
-  cxx_range_for
-  cxx_strong_enums
-  cxx_uniform_initialization
-  # Features that MSVC 18.0 cannot support but in list of Geant4 coding
-  # guidelines - to be required once support for that compiler is dropped.
-  # Version 10.2 is coded without these being required.
-  #cxx_deleted_functions
-  #cxx_generalized_initializers
-  #cxx_constexpr
-  #cxx_inheriting_constructors
-  )
-
-# If a standard higher than 11 has been selected, check that compiler has
-# at least one feature from that standard and append these to the required
-# feature list
-if(GEANT4_BUILD_CXXSTD GREATER 11)
-  if(CMAKE_CXX${GEANT4_BUILD_CXXSTD}_COMPILE_FEATURES)
-    list(APPEND GEANT4_TARGET_COMPILE_FEATURES ${CMAKE_CXX${GEANT4_BUILD_CXXSTD}_COMPILE_FEATURES})
-  else()
-    message(FATAL_ERROR "Geant4 requested to be compiled against C++ standard '${GEANT4_BUILD_CXXSTD}'\nbut detected compiler '${CMAKE_CXX_COMPILER_ID}', version '${CMAKE_CXX_COMPILER_VERSION}'\ndoes not support, or CMake (${CMAKE_VERSION}) is not aware of, any features of that standard.")
-  endif()
-endif()
-
-# - Check for Standard Library Implementation Features
-# Smart pointers are a library implementation feature
-# Hashed containers are a library implementation feature
-# Random numbers are a library implementation feature?
-# An example of where a workaround is needed
-# Rest of concurrency a library implementation feature
-
-# Hold any appropriate compile flag(s) in variable for later export to
-# config files. Needed to support clients using late CMake 2.8 where compile features
-# are not available.
-set(GEANT4_CXXSTD_FLAGS "${CMAKE_CXX${GEANT4_BUILD_CXXSTD}_STANDARD_COMPILE_OPTION}")
-
```

Comment:
This should be _partially_ retained so that the CMake `cxx_std_XY` feature for the required standard is set.

Reply:
Setting the standard to 11 or more should be sufficient to make sure that all those features are available, shouldn't it?

Reply:
No, `CMAKE_CXX_STANDARD` is only part of the picture:

- CMake only knows that a compiler is aware of a given standard, not the exact set of features (at least for 14 and better) supported by that compiler _and_ standard library combination.
- Only compile features can be set as [usage requirements](https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html?#build-specification-and-usage-requirements). That ensures client code using Geant4 is compiled against at the same or newer standard as the install of Geant4 used.

CMake only enumerates the `cxx_std_XY` features now, so we still need that set for the value of `CMAKE_CXX_STANDARD`. Whilst we retain support for C++11, the existing features must be retained. Yes, this means for 17 and better we need to write feature checks, and that's being worked on.

Reply:
Thank you for taking the time in wring these answers :-)
I see... As you said, this is very difficult to maintain, requiring curating language features as they come out. My question is, why are these checks so critical? As far as I can tell, if the feature is missing, the build will fail (instead of the CMake configuration failing); but having a build fail due to unmet requirements is not uncommon anyway. The same would apply for client code. And I'd wager that the Geant4 code base uses C++11 features which are not checked explicitly...

Reply:
Because a configure time check is _way_ faster and clearer than deep in the build stage.

Remember also that we have to balance conflicting requirements across a highly diverse set of users of Geant4. Some want bleeding edge standards/compilers/features, others need stability and reliability. For the former case we can anticipate (as was done for C++11) _optionally_ enabling features for say C++23 so they can be tested, used if available, but not require all users to have a bleeding edge setup.

Reply:
> Because a configure time check is _way_ faster and clearer than deep in the build stage.

Well, sure, I certainly try myself to catch those. But in this particular case, I don't think that this will happen often: the only possibility is that a non-C++11-compliant compiler has been selected by accident. The trade-off is a large amount of maintenance work... And again, I don't know when was the last time that the feature list was updated, but I am sure that some C++11 features have sneaked in without being added to the list. 

> Remember also that we have to balance conflicting requirements across a highly diverse set of users of Geant4. Some want bleeding edge standards/compilers/features, others need stability and reliability. For the former case we can anticipate (as was done for C++11) optionally enabling features for say C++23 so they can be tested, used if available, but not require all users to have a bleeding edge setup.

But are we talking about client code? Because again, setting CMAKE_CXX_STANDARD to 11 does not prevent client code to use whatever experimental compiler features; enabling future features would only be relevant for Geant4 code. And of course those experimental things should never ever go into master (unless we go back to explicitly enabling the new bits with macros, which I don't think is a good idea).

Reply:
O.k., I think we're talking past each other here. 

> Well, sure, I certainly try myself to catch those. But in this particular case, I don't think that this will happen often: the only possibility is that a non-C++11-compliant compiler has been selected by accident. The trade-off is a large amount of maintenance work... And again, I don't know when was the last time that the feature list was updated, but I am sure that some C++11 features have sneaked in without being added to the list.

Still, I'd regard any error picked during compilation related to an unchecked C++XY feature   is a bug - that should be addressed as a configure time check. These updates can be iterative and that makes maintenance easier.

> But are we talking about client code? Because again, setting CMAKE_CXX_STANDARD to 11 does not prevent client code to use whatever experimental compiler features; enabling future features would only be relevant for Geant4 code. And of course those experimental things should never ever go into master (unless we go back to explicitly enabling the new bits with macros, which I don't think is a good idea).

Bottom line is that for Geant4, `CMAKE_CXX_STANDARD` needs to be floating, i.e. selectable as 11 or newer at configure time. Development/testing needs this, and is also there _by explicit request of our largest HEP clients_. There's no cost in making it selectable this way, as the default behaviour is exactly as you describe.

Reply:
> O.k., I think we're talking past each other here.

Yes, unfortunately this platform is not very suitable for this type of discussion, answers get all over the place :-/ My apologies if I have read past some of your points.

> Still, I'd regard any error picked during compilation related to an unchecked C++XY feature is a bug - that should be addressed as a configure time check. These updates can be iterative and that makes maintenance easier.

I do not think that it is the build system's task to catch all of these compiler deficiencies. I understand there is a list of supported compilers, so requiring minimum compiler versions etc would be simpler. This is just my opinion, of course.

> Bottom line is that for Geant4, CMAKE_CXX_STANDARD needs to be floating, i.e. selectable as 11 or newer at configure time. Development/testing needs this, and is also there by explicit request of our largest HEP clients. There's no cost in making it selectable this way, as the default behaviour is exactly as you describe.

My point being that those clients *think* that they need that feature, but they actually don't :-D I mean, I still do not see which current use cases would be made impossible by hard-coding the standard. The cost is certainly not huge but, in my opinion, it makes the standard handling non-standard and more cumbersome than necessary (ie the only thing clients should need to care about is whether their compiler supports Geant4's C++ standard).

In any case, convincing clients to accept this... That is a different story. My point is more like, if you see this tiny improvement of value, it may be possible to push this approach forward when the chance arises, eg when switching standard to C++17.